clients/drpoints/drpoints.cpp

```diff
@@ -0,0 +1,439 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* DrPoints: Basic Block Vector (BBV) Client.
+ *
+ * Given a user-defined instruction interval, computes the BBVs (histogram of BB
+ * frequencies within the interval) of a program execution and outputs them in a .bb file.
+ */
+
+#include "dr_api.h"
+#include "dr_defines.h"
+#include "dr_events.h"
+#include "drmgr.h"
+#include "droption.h"
+#include "drreg.h"
+#include "drvector.h"
+#include "drx.h"
+#include "hashtable.h"
+
+namespace dynamorio {
+namespace drpoints {
+namespace {
+
+using ::dynamorio::droption::bytesize_t;
+using ::dynamorio::droption::DROPTION_SCOPE_CLIENT;
+using ::dynamorio::droption::droption_t;
+
+#define FATAL(...)                       \
+    do {                                 \
+        dr_fprintf(STDERR, __VA_ARGS__); \
+        dr_abort();                      \
+    } while (0)
+
+#define HASH_BITS 13
+
+#define MINSERT instrlist_meta_preinsert
+
+#ifdef X86_64
+#    define INLINE_COUNTER_UPDATE 1
+#else
+// TODO i#7685: We don't have the inlining implementation yet for AARCH64 or 32-bit
+// architectures.
+#endif
+
+static droption_t<bytesize_t> instr_interval(
+    DROPTION_SCOPE_CLIENT, "instr_interval", 100000000 /*=100M instructions*/,
+    "The instruction interval for which to generate BBVs",
+    "Divides the program execution in instruction intervals of the specified size and "
+    "generates BBVs as the BB execution frequency within the interval times the number "
+    "of instructions in the BB. Default is 100M instructions.");
+
+static droption_t<bool>
+    print_to_stdout(DROPTION_SCOPE_CLIENT, "print_to_stdout", false,
+                    "Enables printing the Basic Block Vectors to standard output",
+                    "Also prints to standard output on top of generating the .bbv file. "
+                    "Default is false.");
+
+static droption_t<bool> no_out_bbv_file(
+    DROPTION_SCOPE_CLIENT, "no_out_bbv_file", false,
+    "Disables the generation of the output .bbv file",
+    "Disables the generation of the output .bbv file, but still runs the client. Useful "
+    "for unit tests or paired with -print_to_stdout. Default is false.");
```

Comment:
Where is the file normally written? I don't see any control over that: best to add a parameter for that.